numpy/core/src/multiarray/compiled_base.c

```diff
@@ -795,7 +795,52 @@ static int sequence_to_arrays(PyObject *seq,
     return 0;
 }
 
-/* Inner loop for unravel_index */
+/*
+ * Force-cast an array into intp if it is empty. Fail otherwise.
+ *
+ * op        - Pointer to the array pointer. Will be modified if
+ *             recasting is needed.
+ * paramname - Name to be used for error string.
+ */
+static int _cast_to_intp_or_fail(PyArrayObject **op,
+                                 char *paramname)
+{
+    PyArray_Descr *dtype = PyArray_DescrFromType(NPY_INTP);
+    if (dtype == NULL) {
+        return -1;
+    }
+
+    if (!PyArray_CanCastArrayTo(*op, dtype, NPY_SAME_KIND_CASTING)) {
```

Comment:
Why would I want to call PyArray_FromArray() on arrays that are already fine? Wouldn't that make a copy? I don't need to cast non-empty arrays because the values get casted at the iteration point.

This bugfix is to take care of one special case, nothing else is broken, so why should I muck with that?